### DISCUSSION TITLE
How to nest server components inside client components

### DISCUSSION THREAD

**Discussion by josias-r** [vercel/next.js#43407] [Help] [answered]
Why can I not nest a server component inside a client component (next 13)?

I have the following:
```jsx
<Page>
   <ClientComponent />
</Page>

// ClientComponent.jsx
function ClientComponent() {
   return (
      <ServerComponent />
   )
}
```
This will result in an error:
```
Error: Objects are not valid as a React child (found: [object Promise]). If you meant to render a collection of children, use an array instead.
```

What am I misunderstanding?

**Comment by utg1-tdawg** [comment]
Facing the same issue here. `<ClientComponent />` wraps `<ServerComponent />` that contains async/await. 

TS error:
```
'ServerComponent' cannot be used as a JSX component.
  Its return type 'Promise<Element>' is not a valid JSX element.
```

**Comment by tubbo** [comment]
The actual code for server components isn't actually delivered to the client, so you can't mount it directly as a child of a client component. Have you tried wrapping the server component in a suspense boundary? Just curious, not sure it will work. Kinda goes against the whole philosophy.

**Reply by josias-r** [reply]
Yes wrapping it in suspense does not change a thing.
It goes against the philosophy? I though the philosophy is to have *components*  rendered on the *server* and then sent to the client. 
But from what it seems to be "*all components* rendered on the *server* and then sent to the client".

So I can't have a server component being loaded in by the client.

What I was trying to achieve is a simpel load-more list of data. Where the client can click a buttone which would request a new instance of an async server component rendered on the server.

**Reply by josias-r** [reply]
In the [fundamental docs](https://beta.nextjs.org/docs/rendering/fundamentals) it is already depicted as possible:

<img width="714" alt="Screenshot 2022-11-28 at 10 29 14" src="https://user-images.githubusercontent.com/11424820/204242316-004cec69-9945-4b05-9335-b7dca34d1688.png">

But having a purple `S`  as a child of a blue `C` in that graph seems to actually not work as it is shown

**Comment by utg1-tdawg** [comment]
Client components are supposed to be allowed to wrap server components. (recall how client and server components are interleaved)

I believe the issue is that the wrapped server component contains async logic which causes its return type to be Promise<Element> which isn't allowed in client components.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Two things here. First, Server Components, that use async, as children of anything are a bit of an odd ball for TypeScript, so please use:

```tsx
            {/** @ts-expect-error Server Component */}
            <ServerPokemonCard name={pokemon} />
```

When TypeScript is updated to understand async components, then it'll just complain about the directive being unnecessary and you can safely remove it.

Second, AFAIK, the way to have Server Components as children of Client Components, you ought to use composition:

```tsx
// Wrapper component
"use client";

import type { ReactNode } from "react";
import { SearchPokemon } from "../components/SearchPokemon";

export function Wrapper({ children }: { children: ReactNode }) {
  return (
    <section>
      <h2>Search for a Pokemon</h2>

      <SearchPokemon />

     {children}
    </section>
  );
}

// Page code, notice that this is a SC itself
import { PokemonCard } from "../../components/PokemonCard";

import { ClientWrapper } from "../../components/Wrapper";

export default function Page() {
  return (
    <ClientWrapper>
      {/** @ts-expect-error Server Component */}
      <PokemonCard name="pikachu" />
    </ClientWrapper>
  );
}

// And then PokemonCard is an async component 
```

Read this part carefully... https://beta.nextjs.org/docs/rendering/server-and-client-components#importing-server-components-into-client-components 

The downside is that passing props to server components is not so straight forward. So far I've fallen back to use search params (queries). I am still investigating!

**Reply by josias-r** [reply]
@icyJoseph Interesting, thanks for linking that documentation, I did indeed not see it, it describes exactly the limitation I encountered.

This limitation seems to be quite a big one though. It seems to _completely limit the ability of having interactions based on a client component reflect on the server components_ rendered.

Even if I use the server component as a child, I can still not pass dynamic props to it based on a local state in the client component, since for that case I woul need a function, **but that is not allowed because it is not serializable**:


```tsx
// Wrapper component
"use client";

import type { ReactNode } from "react";
import { SearchPokemon } from "../components/SearchPokemon";

export function Wrapper({ render }: { render: (seachTerm: string) => ReactNode }) {
  const [searchTerm, setSearchTerm] = useState("");
  return (
    <section>
      <h2>Search for a Pokemon</h2>

      <SearchPokemon onSearch={(newTerm)=>{setSearchTerm(newTerm)}} />

     {render(searchTerm)}
    </section>
  );
}

// Page code, notice that this is a SC itself
import { PokemonCard } from "../../components/PokemonCard";

import { ClientWrapper } from "../../components/Wrapper";

export default function Page() {
  return (
    <ClientWrapper render={(search) => (
      <>
        {/** @ts-expect-error Server Component */}
        <PokemonCard name={search} />
      </>
    )} />
  );
}

// And then PokemonCard is an async component 

```

**Reply by icyJoseph** [reply]
Right, that's why I said that, currently, composition is the way to go.

The problem seems to be that Client couples to Server Component (which actually does async work) via props, then we are faced with a boundary. In practice, two computers have to start working together (during development all you have is your machine acting as client and server), and that brings all of the implications of asynchronous work:

- it has an unknown loading time
- it may succeed
- it may fail


I'd personally expect the ability to wrap the SC with a Suspense + Error Boundary to account for the above, but currently it does not work. 

And I suppose one of the biggest issues would be that several client components feeding several server components with props, starts to sound very complex. Rather if every SC would be an entry point into the React tree, which can use cookies, queries, etc, to render data, then we'd manage to control this complexity.

Additionally we ought to think of:

- Props change between renders, and there's ongoing SC fetching, we ought to cancel the ones which are no longer valid
- Props change and the SC component is no longer part of the tree
- DX around what can be passed into the SC boundary... basically serialization expectations


I found this response in the original RFC: https://github.com/reactjs/rfcs/pull/188#issuecomment-750358585, namely this part:

> 

> > Did the team consider allowing for server components to be nested inside of client components? I understand why this isn't allowed because it would be a pretty big foot gun that could reintroduce waterfalls on the client side again. I can think of a few really useful examples of wanting to do this, ex: you might want to render a server component inside of a modal that only opens when the user takes a certain action
> 
> Note you can nest SC into CC as long as the nesting itself happens in another SC above it. But generally saying, there is always some kind of an "entry point" into the SC world, and you're right in some cases it makes sense for this entry point to be "lazy". We shouldn't make this too easy like you said, but technically this capability will definitely exist, including for the use case you mention. Although that use case might also have a more idiomatic solution on the server side.
>
> > Is the suggested way to do something like this have a server component that passes the modal to a client component and add a query param like ?modal=share to the route to signal that a modal should be rendered?
> 
> Yes, something like this might work. Routing is still an open research area, and lazy interfaces like tooltips, modals, etc, will be a part of that research. We need to get Server Context support in first before exploring this.

**Reply by josias-r** [reply]
> I'd personally expect the ability to wrap the SC with a Suspense + Error Boundary to account for the above, but currently it does not work.

Thats exactly what I was hoping for to be able to do. Have a CC with some `useState` that will render a SC with the state as a prop. That SC component could then handle all the different loading-states with an ErrorBoundry and Suspense.

From what I am understanding, at the moment it is only possible to pass client "states" via browser navigation to SC. Meaning SearchParams and Routes. Or am I missing something?

To break it down, anything that can't fit into an URL, cannot be handeled via SC.

So I can have smth simple like
`https://mypage.com?modalOpen=true`, but not `https://mypage.com/list?offset=99` (for a `load more`list). Because with SC we can of course query the next N items with an offset of `99`, but we would also have to load the 99 items again that are supposibly already loaded from by the client.

**Reply by icyJoseph** [reply]
> SearchParams and Routes. Or am I missing something?
> 
> To break it down, anything that can't fit into an URL, cannot be handled via SC.

That's where I am currently at as well.

Not sure about your example though. As much as I like that people are eager to jump into these new concepts, I am also a bit worried about how many seem to want to take this all the way to production, or rather to a point where a beta release can't really take us all. These are still developing technologies. Perhaps in the future more possibilities will open up.

**Reply by josias-r** [reply]
> That's where I am currently at as well.

👍 
> As much as I like that people are eager to jump into these new concepts

Gotta jump into the concepts to find out the limitations it has 😄  And I'm glad I did. This limitation is quite a substantial one, and really makes the whole concept of SC only half as interesting to me. Without the possiblity to have the components dynamically rendered, the whole SC topic to me is just a new improved way of fetching data server-side on the _initial_ request, and not truly component-based data fetching. I could just as well fech all the data from the root page and then pass it down via props with that limitation

**Reply by icyJoseph** [reply]
Well, I did fix this with an api route and the `use` hook, I did not want to show it because I rolled my own cache function, given that React is still preparing theirs, do remember that React is still at 18.0.2:

```tsx
import { Suspense } from "react";
import { ErrorBoundary } from "components/Boundary";
import { ClientPokemonCard } from "components/ClientPokemonCard";
import { PokemonCard } from "components/PokemonCard";
import { SearchPokemon } from "components/SearchPokemon";

export default async function Page({
  searchParams,
}: {
  searchParams: { pokemon?: string };
}) {
  const pokemon = searchParams.pokemon;

  return (
    <section>
      <h2>Search for a Pokemon</h2>

      <SearchPokemon />

      <ErrorBoundary>
        <Suspense fallback={<p>searching...</p>}>
          <div>
            {/** @ts-expect-error Server Component */}
            <PokemonCard name={pokemon} />
          </div>
        </Suspense>
        <Suspense fallback={<p>searching 2...</p>}>
          <div>
            <ClientPokemonCard name={pokemon} />
          </div>
        </Suspense>
      </ErrorBoundary>
    </section>
  );
}

```

And the client version does:

```tsx
export const ClientPokemonCard = ({ name }: { name?: string }) => {
  if (!name) return <p>Search for something</p>;

  const pokemon = use(fetchPokemon(name));

  if ("status" in pokemon && pokemon.status === 404) {
    return <p>{name} is not a pokemon</p>;
  }

  if ("statusText" in pokemon) {
    return <p>Unexpected result searching for {name}</p>;
  }

  return (
    <div>
      <h3>{pokemon.name}</h3>

      <p>{pokemon.order}</p>

      <figure>
        <Image
          src={pokemon.sprites.front_default}
          alt={pokemon.name}
          width={200}
          height={200}
          style={{ imageRendering: "pixelated" }}
        />
      </figure>
    </div>
  );
};
```

Where the trick is in how I made `fetchPokemon` (which itself contains a bunch of delays and other experiments I am doing):

https://user-images.githubusercontent.com/21013447/204373287-3ea1bbf0-5951-4a4a-ba8d-2cc596466050.mov

So in the contrary for me this is all very fascinating and half as more interesting than I initially thought it'd be 😅

**Reply by icyJoseph** [reply]
Don't forget to mark as answered if it has satisfied you for the time being. As said earlier, it is not time to lose sleep over understanding these API's, rather, we should experiment and give feedback on how our experiments go.

**Reply by josias-r** [reply]
Interesting stuff. Im not that deep into everything yet, but what you are showing certainly looks very interesting.

Definetly keeping an I on this, what direction it will take.
